lib/plain/wrappers/wrap.ts

```diff
@@ -16,19 +16,52 @@ export type MarkOptional<BaseType, Keys extends keyof BaseType = keyof BaseType>
     Partial<Pick<BaseType, Keys>>
 
 export type WrapParams = {
-  makeRequest: MakeRequestWithoutUserAgent
+  makeRequest: MakeRequest
   defaults?: DefaultParams
 }
 
-export const wrap = <Params extends {}, Payload extends {}>(
+export type WrapFn<
+  ET extends keyof MRActions,
+  Action extends keyof MRActions[ET],
+  Params = 'params' extends keyof MROpts<ET, Action, false>
+    ? MROpts<ET, Action, false>['params']
+    : undefined,
+  Payload = 'payload' extends keyof MROpts<ET, Action, false>
+    ? MROpts<ET, Action, false>['payload']
+    : undefined,
+  Headers = 'headers' extends keyof MROpts<ET, Action, false>
+    ? MROpts<ET, Action, false>['headers']
+    : undefined,
+  Return = MRReturn<ET, Action>
+> = Params extends undefined
+  ? () => Return
+  : Payload extends undefined
+  ? (params: Params) => Return
+  : Headers extends undefined
+  ? (params: Params, payload: Payload) => Return
+  : (params: Params, payload: Payload, headers: Headers) => Return
+
+export const wrap = <ET extends keyof MRActions, Action extends keyof MRActions[ET]>(
   { makeRequest, defaults }: WrapParams,
-  entityType: string,
-  action: string
-) => {
-  return (params?: Params, payload?: Payload, headers?: Record<string, unknown>): Promise<any> =>
+  entityType: ET,
+  action: Action
+): WrapFn<ET, Action> => {
+  // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
+  // @ts-ignore
```

Comment:
Same same same 😓  Rule of thumb should be to comment `@ts-ignore`s so we know why they are there